File: easy/20-valid_parenthesis.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []  # Hold opening brackets
        all_brackets = {'(':')', '[':']', '{':'}'}  # Dictionary to hold brackets
        for char in s:
            if char in all_brackets: stack.append(char)  # If opening bracket, add to stack
            elif(not stack or char != all_brackets[stack.pop()]): return False  # If stack is empty or brackets dont match, return false
        return not stack  # return true if stack is empty, all brackets matched
                

File: 20-valid_parenthesis.cc
#include<stack>
#include<string>
class Solution {
public:
    bool isValid(std::string s) {
    std::stack<char> opStack;
 
    for(char c : s){
        if(c == '(' || c == '[' || c == '{') opStack.push(c);
        else {
            if(opStack.empty()) return false; // no corresponding open bracket
            char compare = opStack.top();
            opStack.pop();
            if(c==')' && compare!=(c-1)) return false;
            if(c!=')' && compare!=(c-2)) return false;
        }
    }
    if(!opStack.empty()) return false;
    return true;
        
    }
};